capsules/src/ble_advertising_driver.rs

```diff
@@ -457,26 +359,26 @@ impl App {
                 data.as_mut()[PACKET_ADDR_END] = 0xf0;
                 ReturnCode::SUCCESS
             })
-            .unwrap_or_else(|| ReturnCode::ESIZE)
+            .unwrap_or(ReturnCode::FAIL)
     }
 
     fn reset_payload(&mut self) -> ReturnCode {
         match self.process_status {
             Some(BLEState::Advertising(_)) | Some(BLEState::Scanning(_)) => ReturnCode::EBUSY,
             _ => {
-                let res = self.advertisement_buf
+                self.advertisement_buf
                     .as_mut()
                     .map(|data| {
                         for byte in data.as_mut()[PACKET_PAYLOAD_START..PACKET_LENGTH].iter_mut() {
                             *byte = 0x00;
                         }
+
+                        // Reset header length
+                        data.as_mut()[PACKET_HDR_LEN] = EMPTY_PACKET_LEN;
+
                         ReturnCode::SUCCESS
                     })
-                    .unwrap_or_else(|| ReturnCode::EINVAL);
-                if res == ReturnCode::SUCCESS {
-                    self.idx = PACKET_PAYLOAD_START;
-                }
-                res
+                    .unwrap_or(ReturnCode::EINVAL)
             }
         }
     }
```

Comment:
This becomes more or less now otherwise we have to change ```libtock``` because the advertisement buffer is only transmitted to the kernel when ```start_advertising``` is called. We could transmit the new buffer every time it becomes modified though!